auto_nag/topcrash.py

```diff
@@ -0,0 +1,283 @@
+
+        Args:
+            date: the final date
+            duration: the number of days to retrieve the crash data
+
+        Returns:
+            A dictionary where the keys are crash signatures, and the values are
+            list of criteria that the crash signature matches.
+        """
+
+        start_date = lmdutils.get_date(date, duration)
+        end_date = lmdutils.get_date(date)
+        date_range = socorro.SuperSearch.get_search_date(start_date, end_date)
+        self.blocked_signatures = self._fetch_signatures_from_patters(
+            self.signature_block_patterns, date_range
+        )
+
+        data: dict = defaultdict(dict)
+        searches = [
+            socorro.SuperSearch(
+                params=self.__get_params_from_criterion(date_range, criterion),
+                handler=self.__signatures_handler(criterion),
+                handlerdata=data[criterion["name"]],
+            )
+            for criterion in TOP_CRASH_IDENTIFICATION_CRITERIA
+        ]
+
+        for search in searches:
+            search.wait()
+
+        # We merge the results after finishing all queries to avoid race conditions
+        result = {}
+        for _, signatures in data.items():
+            for signature_name, signature_info in signatures.items():
+                if signature_name not in result:
+                    result[signature_name] = [signature_info]
+                else:
+                    result[signature_name].append(signature_info)
+
+        return result
+
+    def __get_params_from_criterion(self, date_range: list, criterion: dict):
+        params = {
+            "product": criterion["product"],
+            "release_channel": criterion["channel"],
+            "process_type": criterion.get("process_type"),
+            "cpu_arch": criterion.get("cpu_arch"),
+            "platform": criterion.get("platform"),
+            "date": date_range,
+            "_aggs.signature": [
+                "startup_crash",
+            ],
+            "_results_number": 0,
+            "_facets_size": (
+                criterion.get("tc_startup_limit", criterion["tc_limit"])
+                # Because of the limitation in https://bugzilla.mozilla.org/show_bug.cgi?id=1257376#c9,
+                # we cannot ignore the generic signatures in the Socorro side, thus we ignore them
+                # in the client side. We add here the maximum number of signatures that could be
+                # ignored to stay in the safe side.
+                + len(self.blocked_signatures)
+            ),
+        }
+        return params
+
+    @staticmethod
+    def __is_startup_crash(signature: dict):
+        return any(
+            startup["term"] == "T" for startup in signature["facets"]["startup_crash"]
+        )
+
+    def __signatures_handler(self, criterion: dict):
+        def handler(search_resp: dict, data: dict):
+            """
+            Handle and merge crash signatures form different quires.
```

Comment:
Typos

Reply:
Fixed in dae46bd16648cbdefa642a0c73f98de6cbd00d7d